core/transfer/src/main/java/org/eclipse/dataspaceconnector/transfer/core/commandqueue/BoundedCommandQueue.java

```diff
@@ -0,0 +1,89 @@
+package org.eclipse.dataspaceconnector.transfer.core.commandqueue;
+
+import org.eclipse.dataspaceconnector.spi.transfer.command.Command;
+import org.eclipse.dataspaceconnector.spi.transfer.command.CommandQueue;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+/**
+ * In-memory implementation of the {@link CommandQueue} that is backed by a bounded queue. This means that attempting
+ * to add elements to an already full queue will fail and raise an {@link IllegalStateException}.
+ * <p>
+ * This queue is threadsafe, so every operation on it can be assumed atomic.
+ */
+public class BoundedCommandQueue implements CommandQueue {
+    private final ArrayBlockingQueue<Command> queue;
+    private final ReentrantLock lock;
+
+    public BoundedCommandQueue(int bound) {
+        lock = new ReentrantLock();
```

Comment:
the lock is not needed, as the `ArrayBlockingQueue` already implements a lock mechanism

Reply:
the `drainTo(Collection<? super E> c, int maxElements)` method could be used instead, it's part of the `BlockingQueue` interface